src/Microsoft.AspNetCore.Razor.Evolution/Legacy/RazorParser.cs

```diff
@@ -25,17 +24,19 @@ public RazorParser(RazorParserOptions options)
 
         public RazorParserOptions Options { get; }
 
-        public virtual RazorSyntaxTree Parse(TextReader input) => Parse(input.ReadToEnd());
-
-        public virtual RazorSyntaxTree Parse(string input) => Parse(((ITextDocument)new SeekableTextReader(input)));
+        public virtual RazorSyntaxTree Parse(RazorSourceDocument source)
+        {
+            if (source == null)
+            {
+                throw new ArgumentNullException(nameof(source));
+            }
 
-        public virtual RazorSyntaxTree Parse(char[] input) => Parse(((ITextDocument)new SeekableTextReader(input)));
+            var chars = new char[source.Length];
+            source.CopyTo(0, chars, 0, source.Length);
 
-        public virtual RazorSyntaxTree Parse(ITextDocument input) => ParseCore(input);
+            var reader = new SeekableTextReader(chars);
```

Comment:
Going to send another PR when we implement this on top of `RazorSourceDocument` instead of making a copy